advanced/demos/python_GUI.py

```diff
@@ -0,0 +1,318 @@
+        return Mesh(v, m.edges, m.faces, m.face_vectors)
+
+    #used as a callback for the virtual robot to keep bases alligned
+    def set_base_rotation(self, feedback):
+        self.base_rotation = feedback.orientation
+
+    # this functions uses forward kinematics to transform base meshes into the position of the robot component
+    # both vertices and face_vectors are transformed so that the shading is consistent with the model position
+    # this function is called by the feedback callback (this is assigned in the app class)
+    def generate_3d_mesh(self, feedback):
+        self.mesh = []  # creates empty list of meshes
+        angles = feedback.position  # get the angles of the actuators
+        base_rotation = Quaternion(feedback.orientation[0]).transformation_matrix  # determines rotation of the base
+        transforms = self.model.get_forward_kinematics('output', angles)
+        # adds an identity matrix to the list of transforms for consistent indexing (otherwise the first transform would
+        # refer to the position of the second component
+        transforms.insert(0, numpy.matrix([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]]))
+        # multiplies base rotation and the kinematic transforms (unless there are nan values in the quaternion)
+        if not numpy.isnan(numpy.array(feedback.orientation[0],dtype=numpy.float64)).any():
+            transforms = [numpy.dot(base_rotation, transforms[i]) for i in range(len(self.components))]
+        for i in range(len(self.components)):
+            if self.components[i][0] == 'bracket':  # no model for the bracket
+                continue
+            m = Mesh([], [], [], [])
+            bm = Mesh(base_meshes[self.components[i][0]].vertices, base_meshes[self.components[i][0]].edges,
+                      base_meshes[self.components[i][0]].faces, base_meshes[self.components[i][0]].face_vectors)
+            if self.components[i][0] == 'link':
+                bm = self.parse_mesh(bm, [.5, self.components[i][2]])
+            if self.components[i][0] == 'actuator':
+                bm = self.parse_mesh(bm, [.75])
+            m.edges = bm.edges  # edges and faces are the same for base and transformed meshes
+            m.faces = bm.faces
+            for j in range(len(bm.vertices)):
+                # append one to end of vertex to make matrix multiplication possible (transformation matrices are 4x4
+                v = [bm.vertices[j][0], bm.vertices[j][1], bm.vertices[j][2], 1]
+                vT = numpy.dot(transforms[i], v)  # apply transform to vertex
+                m.vertices.append(vT.A1[0:3])  # add transformed vertex to mesh
+            for j in range(len(bm.face_vectors)):  # similar but for face vectors
+                normal = bm.face_vectors[j]
+                TRot = transforms[i][numpy.ix_([0, 1, 2], [0, 1, 2])]
+                nT = numpy.dot(TRot, normal)
+                m.face_vectors.append(nT.A1)
+            self.mesh.append(m)
+
+'''
```

Comment:
same comment above re: comment. Sorry to be nitpicky about this, but I'm not sure Python is smart enough to associate this comment with the `App` class. I only really bring this up just so help docs (i.e., typing in `App?` in `ipython3`) in python get read in properly, and so documentation (if we ever generate any) can be easily generated.